### PR TITLE
Feature #177161074-#177161070 – Cell type wheel and cell type heatmap endpoints

### PR DESCRIPTION
**Pull request by alfonsomunozpomer**
**Notice this is a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/)**.

Regarding the controllers, there are two endpoints: one for the Cell Type Wheel and another one for the Multiexperiment Cell Type Heatmap.

The envisioned workflow is:
1. The JSP view instantiates the Cell Type Wheel Multiexperiment Cell Type Heatmap wrapped with [`atlas-react-fetch-loader`](https://github.com/ebi-gene-expression-group/atlas-components/tree/master/packages/atlas-react-fetch-loader), passing a search term as a prop.
2. The component makes a fetch request to the cell type wheel endpoint, gets a JSON payload and displays the wheel (the heatmap at that point shows a placeholder message asking the user to choose a wheel outer slice).
3. When the user clicks on a cell type in the wheel the component makes a request to the heatmap endpoint and shows the corresponding chart to the right.

Special mention goes to `CellTypeWheelDao` class, with a recursive general solution to extract facets from a Solr response with nested facet terms.

Unless there is some bugs hiding in plain sight, the only things missing are tests: for the Highcharts sunburst adapter, for the services and finally web integration tests for the controllers.

### CODE REVIEW

File: app/src/main/java/uk/ac/ebi/atlas/search/metadata/CellTypeWheelDao.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+import static uk.ac.ebi.atlas.solr.cloud.collections.SingleCellAnalyticsCollectionProxy.CTW_ORGANISM;
+import static uk.ac.ebi.atlas.solr.cloud.collections.SingleCellAnalyticsCollectionProxy.CTW_ORGANISM_PART;
+import static uk.ac.ebi.atlas.solr.cloud.collections.SingleCellAnalyticsCollectionProxy.EXPERIMENT_ACCESSION;
+import static uk.ac.ebi.atlas.solr.cloud.collections.SingleCellAnalyticsCollectionProxy.ONTOLOGY_ANNOTATION_ANCESTORS_LABELS;
+import static uk.ac.ebi.atlas.solr.cloud.collections.SingleCellAnalyticsCollectionProxy.ONTOLOGY_ANNOTATION_LABEL;
+import static uk.ac.ebi.atlas.solr.cloud.collections.SingleCellAnalyticsCollectionProxy.ONTOLOGY_ANNOTATION_PARENT_LABELS;
+import static uk.ac.ebi.atlas.solr.cloud.collections.SingleCellAnalyticsCollectionProxy.ONTOLOGY_ANNOTATION_PART_OF_LABELS;
+import static uk.ac.ebi.atlas.solr.cloud.collections.SingleCellAnalyticsCollectionProxy.ONTOLOGY_ANNOTATION_SYNONYMS;
+
+@Component
```
Comment (upendrakumbham) [inline app/src/main/java/uk/ac/ebi/atlas/search/metadata/CellTypeWheelDao.java:26]: @Repository suits more than @Component here as this is contacting data repository.

File: app/src/test/java/uk/ac/ebi/atlas/experimentpage/markergenes/JsonMarkerGenesControllerWIT.java [hunk 2]
```diff
@@ -78,12 +79,13 @@ void setUp() {
         this.mockMvc = MockMvcBuilders.webAppContextSetup(this.wac).build();
     }
 
-    @Test
+    @RepeatedTest(10)
```
Comment (upendrakumbham) [inline app/src/test/java/uk/ac/ebi/atlas/experimentpage/markergenes/JsonMarkerGenesControllerWIT.java:82]: Changed this from @Test to @RepeatedTest - I know this test is a flaky one that fails very randomly or any other reason for this test to run 10 times?

File: app/src/test/java/uk/ac/ebi/atlas/search/metadata/MultiexperimentCellTypeMarkerGenesDaoIT.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        populator.setScripts(
+                new ClassPathResource("fixtures/202108/scxa_cell_group_marker_gene_stats-delete.sql"),
+                new ClassPathResource("fixtures/202108/scxa_cell_group_marker_genes-delete.sql"),
+                new ClassPathResource("fixtures/202108/scxa_cell_group-delete.sql"),
+                new ClassPathResource("fixtures/202108/experiment-delete.sql"));
+        populator.execute(dataSource);
+    }
+
+    @Test
+    void foobar() {
```
[the remark this answers is not in the document]
Comment (alfonsomunozpomer) [reply app/src/test/java/uk/ac/ebi/atlas/search/metadata/MultiexperimentCellTypeMarkerGenesDaoIT.java]: Done!

File: app/src/main/java/uk/ac/ebi/atlas/search/metadata/JsonCellTypeWheelController.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import org.springframework.http.MediaType;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RestController;
+import uk.ac.ebi.atlas.controllers.JsonExceptionHandlingController;
+
+import static uk.ac.ebi.atlas.utils.GsonProvider.GSON;
+
+@RestController
+public class JsonCellTypeWheelController extends JsonExceptionHandlingController {
```
[the remark this answers is not in the document]
Comment (upendrakumbham) [reply app/src/main/java/uk/ac/ebi/atlas/search/metadata/JsonCellTypeWheelController.java:12]: Does it improve readability in this context? Sorry, if it is silly question :).

File: docker/docker-compose-gradle.yml [hunk 5]
```diff
@@ -12,7 +12,11 @@ services:
     volumes:
       - ..:/root/project
       - ./packages:/root/.m2
-      - $ATLAS_DATA_PATH:/root/scxa/integration-test-data
+      - /home/alf/gradle-ro-dep-cache:/gradle-ro-dep-cache
```
[the remark this answers is not in the document]
Comment (alfonsomunozpomer) [reply docker/docker-compose-gradle.yml:15]: Oh, yes, please do. I guess an env variable such as `$GRADLE_RO_DEP_CACHE` is much better.

File: app/src/main/java/uk/ac/ebi/atlas/search/metadata/JsonCellTypeWheelController.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (alfonsomunozpomer) [reply app/src/main/java/uk/ac/ebi/atlas/search/metadata/JsonCellTypeWheelController.java:12]: It’s not silly. This is more of a convention that goes back a long time: every controller that extends `JsonExceptionHandlingController` has the `Json` prefix. It’s not a huge help when you read the code of a class in isolation, but it’s handy when you move around the project tree (think of it as things like suffixes such as `Test`, `IT`, `Dao`, etc.).

[the remark this answers is not in the document]
Comment (upendrakumbham) [reply app/src/main/java/uk/ac/ebi/atlas/search/metadata/JsonCellTypeWheelController.java:12]: Thanks for the clarification @alfonsomunozpomer.

File: app/src/main/java/uk/ac/ebi/atlas/search/metadata/CellTypeWheelDao.java (hunk 1 above)
Comment (haideriqbal) [reply app/src/main/java/uk/ac/ebi/atlas/search/metadata/CellTypeWheelDao.java:26]: I do not have any strong opinion in this case. I believe `@Component` annotation would work in this case considering the argument which Alfonso shared. Also, looking here https://www.baeldung.com/spring-component-repository-service @Repository’s job is to catch persistence-specific exceptions and re-throw them as one of Spring’s unified unchecked exceptions. So again, it really depends.

**Comment by alfonsomunozpomer** [comment]
Merging after succesful run on my local environment.

**Pull request merged**